src/librustc_mir/interpret/memory.rs

```diff
@@ -806,14 +811,29 @@ impl<'mir, 'tcx, M: Machine<'mir, 'tcx>> Memory<'mir, 'tcx, M> {
 
         let tcx = self.tcx.tcx;
 
-        // This checks relocation edges on the src.
-        let src_bytes = self.get(src.alloc_id)?
-            .get_bytes_with_undef_and_ptr(&tcx, src, size)?
-            .as_ptr();
         let dest_bytes = self.get_mut(dest.alloc_id)?
             .get_bytes_mut(&tcx, dest, size * length)?
             .as_mut_ptr();
 
+        // This checks relocation edges on the src.
+        let src_bytes = self.get(src.alloc_id)?
+            .get_bytes_with_undef_and_ptr(&tcx, src, size)?;
+
+        let src_bytes = match src_bytes {
+            BytesOrElidedUndef::MaybeBytes(bytes) => bytes.as_ptr(),
+            BytesOrElidedUndef::ElidedUndef => {
+                // All bytes are undef. There are no source bytes to copy from but if they were
+                // allocated right now they might be all zeroes.
+                let target_alloc = self.get_mut(dest.alloc_id)?.alloc.as_mut().unwrap();
```

Comment:
why `unwrap` and not `force_alloc`?

Reply:
actually, if `alloc.is_none()` you don't need to do anything here.

Reply:
Doing nothing if `alloc.is_none()` works out well with the restoring the previous order, in that `dest` will not needlessly allocate the bytes from `get_mut` because that point is never reached. Makes reasoning about `src == dest` easier as well.